lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -0,0 +1,251 @@
+        else if (connect.getSrc() == instanceResult)
+          driver = connect.getDest();
+        // Downward reference constraint ensures that it must be a block
+        // argument.
+        auto blockArg = driver.cast<BlockArgument>();
+        // Record the remote op from which the dataflow is established to this
+        // module port.
+        refPortToXmrMap[std::make_pair(
+            instanceResult.getDefiningOp()->getParentOfType<FModuleOp>(),
+            blockArg.getArgNumber())] = remoteOp;
+      } else if (auto getOp = dyn_cast<RefResolveOp>(user)) {
+        // The source of the dataflow for this RefResolveOp is established. So
+        // replace the RefResolveOp with the coresponding VerbatimExpr to
+        // generate the XMR.
+        SmallVector<Attribute> xmrHierPath(
+            remoteRefToXMRsymVec[remoteOp].rbegin(),
+            remoteRefToXMRsymVec[remoteOp].rend());
+        SmallString<128> xmrString;
+        for (unsigned id = 0, e = xmrHierPath.size(); id < e; ++id) {
+          ("{{" + Twine(id) + "}}").toVector(xmrString);
+          if (id < e - 1)
+            xmrString += '.';
+        }
+        ImplicitLocOpBuilder builder(getOp.getLoc(), getOp);
+        auto xmrVerbatim = builder.create<VerbatimExprOp>(
+            getOp.getType().cast<FIRRTLType>(), xmrString, ValueRange{},
+            xmrHierPath);
+        getOp.getResult().replaceAllUsesWith(xmrVerbatim);
+        LLVM_DEBUG(llvm::dbgs() << "\n The get op:" << getOp
+                                << "\n is connected to :" << remoteOp;
+                   for (auto attr
+                        : remoteRefToXMRsymVec[remoteOp]) llvm::dbgs()
+                   << ":" << attr;);
+      } else
+        user->emitError("unhandled user of reference port");
+      // This reference port will be removed, so remove all its users also.
+      opsToRemove.push_back(user);
+    }
+  }
+
+  bool handleRemoteReference(Operation *op, FModuleOp module) {
+    Value xmrPort, xmrResult;
+    if (auto xmrEnd = dyn_cast<RefSendOp>(op)) {
+      xmrPort = xmrEnd.getRef();
+      xmrResult = xmrEnd.getResult();
+    } else // Handle RefRecvOp when its implemented.
+      return false;
+    LLVM_DEBUG(llvm::errs()
+               << "\n xmr get/end drives :"
+               << "\n for xmr get/end op :" << op << "\n xmr port" << xmrPort);
+    // Get a reference to the actual signal to which the XMR will be generated.
+    auto xmrDefOp = getInnerRefTo(xmrResult.getDefiningOp());
+    remoteRefToXMRsymVec[op] = {xmrDefOp};
+
+    // The downward reference constraint ensures that xmrPort must be a
+    // BlockArg.
+    auto blockArg = xmrPort.cast<BlockArgument>();
+    // Record the remote reference op, that this module port refers to.
+    refPortToXmrMap[std::make_pair(module, blockArg.getArgNumber())] = op;
+    opsToRemove.push_back(op);
+    return true;
+  }
+
+  bool handleInstanceOp(Operation *op, FModuleOp module) {
```

Comment:
This might be cleaner as `handleInstanceOp(InstanceOp op, ...)` and using a `walk` or `TypeSwitch` in `runOnOperation`. For this specific function, it seems a bit weird to move the check into the body when it clearly can only work on `InstanceOp`s.